app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt

```diff
@@ -37,66 +45,113 @@ class MessageNotificationWorker(
-
-            // Generate the processed Message action
-            val processedAction = nimbusMessagingController.processMessageActionToUri(nextMessage)
-            val actionIntent = Intent(Intent.ACTION_VIEW, processedAction)
+            nimbusMessagingController.onMessageDisplayed(updatedMessage)
 
             NotificationManagerCompat.from(context).notify(
                 MESSAGE_TAG,
-                SharedIdsHelper.getNextIdForTag(context, nextMessage.id),
-                buildNotification(nextMessage, actionIntent),
+                SharedIdsHelper.getNextIdForTag(context, updatedMessage.id),
+                buildNotification(
+                    this,
+                    context,
+                    nimbusMessagingController,
+                    updatedMessage,
+                ),
             )
         }
 
         return Result.success()
     }
 
-    private fun Message.shouldDisplayMessage() = metadata.displayCount == 0
-
-    private fun buildNotification(message: Message, intent: Intent): Notification {
-        with(applicationContext) {
-            val pendingIntent = PendingIntent.getActivity(
-                this,
-                SharedIdsHelper.getNextIdForTag(this, NOTIFICATION_PENDING_INTENT_TAG),
-                intent,
-                IntentUtils.defaultIntentPendingFlags,
+    private suspend fun buildNotification(
+        coroutineScope: CoroutineScope,
+        context: Context,
+        nimbusMessagingController: NimbusMessagingController,
+        message: Message,
+    ): Notification {
+        val onClickPendingIntent =
+            createOnClickPendingIntent(
+                coroutineScope,
+                context,
+                nimbusMessagingController,
+                message.id,
             )
+        val onDismissPendingIntent = createOnDismissPendingIntent(context, message)
+
+        return createBaseNotification(
+            context,
+            MARKETING_CHANNEL_ID,
+            message.data.title,
+            message.data.text,
+            onClickPendingIntent,
+            onDismissPendingIntent,
+        )
+    }
 
-            return createBaseNotification(
-                this,
-                MARKETING_CHANNEL_ID,
-                message.data.title,
-                message.data.text,
-                pendingIntent,
-            )
-        }
+    private suspend fun createOnClickPendingIntent(
+        coroutineScope: CoroutineScope,
+        context: Context,
+        nimbusMessagingController: NimbusMessagingController,
+        messageId: String,
+    ): PendingIntent? = withContext(coroutineScope.coroutineContext) {
+        val message =
+            nimbusMessagingController.getMessage(messageId) ?: return@withContext null
+
+        // Update message as 'clicked'.
+        nimbusMessagingController.onMessageClicked(message.metadata)
+
+        val intent = nimbusMessagingController.getIntentForMessageAction(message.action)
```

Comment:
Both of these calls need to be done _after_ the user has clicked on the message.

I wonder if we can use the `IntentReceiverActivity` with a notification specific `IntentProcessor`?

I think we need advice from @Amejia481!

Reply:
I guess it depends how we want to handle a notification click, but all the actions that we are getting are from nimbus are being handled on [HomeDeepLinkIntentProcessor](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/home/intent/HomeDeepLinkIntentProcessor.kt)  :) 

See `toDeepLinkSchemeUri`
https://github.com/mozilla-mobile/fenix/blob/249ddef570c14ea748524fadceb9d55446e28410/app/src/main/java/org/mozilla/fenix/gleanplumb/NimbusMessagingController.kt#L101-L111